# (실습) 넘파이 어레이

**주의사항**

- `None`은 적절한 표현식으로 대체한다.
- `Pass`는 적절한 명령문으로 대체한다.

## numpy 모듈

**문제 1**

아래 코드를 설명하라.

In [1]:
import numpy as np

`코드설명`: 파이썬 내장함수인 Numpy를 불러오고, 해당 함수를 불러올 때 np로 호출한다.

아래 코드는 수정하지 마세요!!!

In [2]:
np.__version__

'2.0.2'

**문제 2**

아래 코드의 기능을 설명하라

In [3]:
np.random.seed(42)

`코드설명`: 난수를 씨드번호 42번에 저장한다. 42번을 불러오게 되면 항상 난수가 출력된다.

**문제 3**

부동소수점으로 구성된 어레이의 항목은 기본적으로 소수점 이하 8자리까지 출력한다.

In [4]:
Z = np.random.random((40,40))
Z

array([[0.37454012, 0.95071431, 0.73199394, ..., 0.09767211, 0.68423303,
        0.44015249],
       [0.12203823, 0.49517691, 0.03438852, ..., 0.07404465, 0.35846573,
        0.11586906],
       [0.86310343, 0.62329813, 0.33089802, ..., 0.18657006, 0.892559  ,
        0.53934224],
       ...,
       [0.81783422, 0.12020906, 0.5444891 , ..., 0.89803055, 0.2434823 ,
        0.92703454],
       [0.06026739, 0.93443603, 0.35162269, ..., 0.41102847, 0.8398613 ,
        0.90002312],
       [0.35342138, 0.23687056, 0.78052552, ..., 0.07941937, 0.08565825,
        0.89419087]])

`np.set_printoptions()` 의 기능을 다음 두 개의 예를 이용하여 설명하라.

In [5]:
np.set_printoptions(precision=4)

In [6]:
Z

array([[0.3745, 0.9507, 0.732 , ..., 0.0977, 0.6842, 0.4402],
       [0.122 , 0.4952, 0.0344, ..., 0.074 , 0.3585, 0.1159],
       [0.8631, 0.6233, 0.3309, ..., 0.1866, 0.8926, 0.5393],
       ...,
       [0.8178, 0.1202, 0.5445, ..., 0.898 , 0.2435, 0.927 ],
       [0.0603, 0.9344, 0.3516, ..., 0.411 , 0.8399, 0.9   ],
       [0.3534, 0.2369, 0.7805, ..., 0.0794, 0.0857, 0.8942]])

In [7]:
np.set_printoptions(precision=6)

In [8]:
Z

array([[0.37454 , 0.950714, 0.731994, ..., 0.097672, 0.684233, 0.440152],
       [0.122038, 0.495177, 0.034389, ..., 0.074045, 0.358466, 0.115869],
       [0.863103, 0.623298, 0.330898, ..., 0.18657 , 0.892559, 0.539342],
       ...,
       [0.817834, 0.120209, 0.544489, ..., 0.898031, 0.243482, 0.927035],
       [0.060267, 0.934436, 0.351623, ..., 0.411028, 0.839861, 0.900023],
       [0.353421, 0.236871, 0.780526, ..., 0.079419, 0.085658, 0.894191]])

`코드설명`: 해당 함수의 기능은 소숫점 자리수를 표현한다. precision에 넣는 값이 소숫점 자리수가 된다.

아래 코드는 기본 옵션값으로 되돌린다. 수정하지 마세요.

In [9]:
np.set_printoptions(precision=8)

## 1차원 어레이

**문제 1**

크기가 10이고 모든 항목이 0인 1차원 어레이를
변수 `Z`가 가리키도록 하라.
단, `np.zeros()` 함수를 반드시 이용한다.

In [10]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros(10)

아래 코드는 수정하지 마세요!!!

In [11]:
assert (Z == np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])).all()

**문제 2**

변수 `Z`가 10에서 49까지의 정수 중에서 짝수로만 구성된 1차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수를 반드시 이용한다.

In [12]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(10,49,2)

아래 코드는 수정하지 마세요!!!

In [13]:
assert (Z == np.array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
                        44, 46, 48])).all()

**문제 3**

변수 `Z`가 아래와 같이 선언되었다.

In [14]:
Z = np.array([2, 3, 1, 0, 4])
Z

array([2, 3, 1, 0, 4])

변수 `Z`의 항목을 정렬하라.
단, sort() 메서드를 활용한다.

In [15]:
# None을 적절한 표현식으로 대체하라.
Z = np.sort(Z)

아래 코드는 수정하지 마세요!!!

In [16]:
assert (Z == np.arange(5)).all()

**문제 4**

변수 `Z`가 크기 30이며 무작위로 생성된 1차원 어레이를 가리키도록 하라.
또한 포함된 항목들의 평균값을 계산하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
`np.random.random()` 함수, `mean()` 어레이 메서드를 반드시 이용한다.

In [17]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(30)  # 0~1 사이의 난수 생성, 균등하게 분포
Z_mean = Z.mean()

아래 코드는 수정하지 마세요!!!

In [18]:
assert (Z.shape == (30,))
print(f"1차원 어레이 평균값: {Z_mean}")

1차원 어레이 평균값: 0.4618868095723934


**문제 5**

1차원 어레이가 다음과 같이 주어졌다.

In [19]:
Z = np.arange(1, 4)

(1) 아래 표현식의 계산 결과를 설명하라.

In [20]:
Z*Z

array([1, 4, 9])

`코드설명`: 현재 Z = ([1,2,3]) 이고, Z*Z는 각 행과 열이 동일한 위치의 값끼리의 곱셈으로 나온다.

(2) 아래 표현식의 계산 결과를 설명하라.

In [21]:
Z**Z

array([ 1,  4, 27])

`코드설명`: 현재 Z = ([1,2,3]) 이고, Z**Z는 각 행과 열이 동일한 위치의 값끼리의 제곱으로 나온다.

(3) 아래 표현식의 계산 결과를 설명하라.

In [22]:
Z < -Z

array([False, False, False])

`코드설명`: 각 행과 열이 동일한 위치의 값을 비교한 결과를 반환한다.

(4) 아래 표현식의 계산 결과를 설명하라.

In [23]:
Z/1/1

array([1., 2., 3.])

`코드설명`: Z = ([1,2,3]) 이고 각 위치에 대해 1로 나눈 결과를 반환한다.

**문제 6**

변수 `Z`가 아래 모양의 어레이를 가리키도록 하라.
단, `np.linspace()` 함수와 `endpoint` 키워드 인자를 이용한다.

```
[0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```

In [24]:
# None을 적절한 표현식으로 대체하라.
Z = np.linspace(0, 1, 10, endpoint=False)

# linspace는 시작과 끝을 포함하는데 endpoint=False로 설정하면 끝을 포함하지 않는다.
# 보통 곡선 그래프를 그릴 때 사용한다

아래 코드는 수정하지 마세요!!!

In [25]:
assert np.isclose(Z, np.array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])).all()

**문제 7**

아래 코드의 실행결과를 설명하라.

In [26]:
Z = np.zeros((10,10))
print("%d bytes" % (Z.size * Z.itemsize))

800 bytes


`코드설명`: Z는 10열 10행이고 0의 값으로 채워져 있다. 따라서 Z 행렬의 크기는 10x10=100이고 np.zeros의 기본 dtype은 float(64)이기 때문에 64비트, 즉 8바이트의 크기를 가진다.

**문제 8**

다음 리스트에서 0이 아닌 항목들의 인덱스로 구성된 어레이를 한줄 코드로 생성하라.
단. `np.nonzero()` 함수를 활용한다.

In [27]:
Z = [1, 2, 0, 0, 4, 0]

In [28]:
# None을 적절한 표현식으로 대체하라.

non_zeros = np.nonzero(Z)

아래 코드는 수정하지 마세요!!!

In [29]:
assert (non_zeros == np.array([0, 1, 4])).all()

**문제 9**

두 개의 1차원 어레이가 다음과 같이 주어졌다.

In [30]:
A = np.ones(3)*1
B = np.ones(3)*2

아래 코드의 실행 결과를 설명하라.

In [31]:
(A+B)*(-A/2)

array([-1.5, -1.5, -1.5])

`코드설명`: A = ([1,1,1]), B = ([2,2,2]) 이기 때문에 (A+B)=([3,3,3]), (-A/2)=([-0.5,-0.5,-0.5]), (A+B)*(-A/2)=([-1.5, -1.5, -1.5]) 이다.

In [32]:
np.add(A,B,out=B)

array([3., 3., 3.])

`코드설명`: A + B 의 결괏값을 B에 저장

In [33]:
np.divide(A,2,out=A)

array([0.5, 0.5, 0.5])

`코드설명`: A / 2 의 결괏값을 A에 저장

In [34]:
np.negative(A,out=A)

array([-0.5, -0.5, -0.5])

`코드설명`: A의 모든 요소에 음수 연산을 적용 후 A에 저장 (지금 A에는 ([0.5, 0.5, 0.5])가 저장이 되어있다)

In [35]:
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

`코드설명`: AxB의 결괏값을 A에 저장

## 2차원 어레이

**문제 1**

변수 `Z`가 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [36]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(25)
Z = np.reshape(Z, (5,5))
print(Z)

[[0.75616333 0.34757152 0.66491172 0.79544996 0.92717782]
 [0.23464208 0.39931592 0.15241601 0.9924835  0.92700097]
 [0.53995713 0.84203329 0.52095798 0.62358569 0.08912443]
 [0.75527042 0.12771348 0.82606763 0.78202809 0.7087447 ]
 [0.03616038 0.30312836 0.26311257 0.36013641 0.08764275]]


아래 코드는 수정하지 마세요!!!

In [37]:
assert (Z.shape == (5, 5))
assert (Z.dtype == np.float64)

**문제 2**

변수 `Z`가 0에서 8까지의 정수로 이루어진 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `reshape()` 어레이 메서드를 반드시 이용한다.

In [38]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(9)
Z = Z.reshape(3,3)
print(Z)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


아래 코드는 수정하지 마세요!!!

In [39]:
assert (Z == np.array([[0, 1, 2],
                       [3, 4, 5],
                       [6, 7, 8]])).all()

**문제 3**

변수 `Z`가 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, 2차원 어레이의 대각선 상은 1이고 나머지는 0이어야 한다.
또한 `np.eye()` 함수를 반드시 이용한다.

In [40]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)
print(Z)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


아래 코드는 수정하지 마세요!!!

In [41]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all

**문제 4**

변수 `Z`가 아래와 같은 (5,5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `np.diag()` 함수의 `k` 키워드 인자를 반드시 활용한다.

```
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
```

In [42]:
# None을 적절한 표현식으로 대체하라.
Z = np.array(25)
Z = np.diag(np.arange(1, 5), k=-1)
print(Z)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


아래 코드는 수정하지 마세요!!!

In [43]:
assert (Z == np.array([[0, 0, 0, 0, 0],
                        [1, 0, 0, 0, 0],
                        [0, 2, 0, 0, 0],
                        [0, 0, 3, 0, 0],
                        [0, 0, 0, 4, 0]])).all()

**문제 5**

변수 `Z`가 아래와 같은 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.zeros()` 함수, `np.arange()` 함수, 2차원 어레이 덧셈을 이용한다.

```
[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]
```

In [44]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros((5,5)) + np.arange(5)

아래 코드는 수정하지 마세요!!!

In [45]:
assert (Z == np.array([[0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.]])).all()

**문제 6**

아래 3x3 단위 행렬에 해당하는 2차원 어레이를 한줄 코드로 생성하라.
단, `np.eye()` 함수를 이용한다.

```
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

In [46]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [47]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all()

**문제 7**

무작위 수로 구성된 (5, 5) 모양의 2차원 어레이가 다음과 같다.

In [48]:
# 지정한 범위로 값을 조정하는 것: 정규화 (0~1 사이의 값으로 조정이 default)
# 대원칙: 인공지능은 학습 전에 데이터 정규화가 되어있어야 한다.

Z = np.random.random((5, 5))

변수 `Z`가 가리키는 2차원 어레이에 대해 min-max 크기조정(정규화)을 적용하라.

min-max 크기조정(정규화)은 항목의 모든 값을 아래 식을 통해 0과 1 사이의 부동소수점으로 변환한다. $max$와 $min$은 각각 최대 항목과 최소 항목을 가리킨다.

$$
\frac{x-min}{max - min}
$$

In [49]:
# None을 적절한 표현식으로 대체하라.
Z = (Z-Z.min())/(Z.max()-Z.min())

아래 코드는 수정하지 마세요!!!

In [50]:
assert (Z.max() == 1 and Z.min() == 0)

## 3차원 어레이

**문제 1**

(1) 변수 `Z`가 (3, 3, 3) 모양의 3차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [51]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((3,3,3))

아래 코드는 수정하지 마세요!!!

In [52]:
assert (Z.shape == (3, 3, 3))
assert (Z.dtype == 'f8')

(2) `Z`에 포함된 항목들 중에서 최댓값과 최솟값의 차이를 구하라.
단, `maximum`, `mininum` 두 변수가 각각 최댓값과 최솟값을 가리키도록 하며,
`max()`와 `min()` 어레이 메서드를 이용한다.

In [53]:
# None을 적절한 표현식으로 대체하라.
maximum = Z.max()
minimum = Z.min()

아래 코드는 수정하지 마세요!!!

In [54]:
print(f"최댓값과 최솟값의 차이: {maximum - minimum}")

최댓값과 최솟값의 차이: 0.9497982939104279


## `dtype` 종류

**문제 1**

아래 두 코드의 실행결과를 설명하라.

In [55]:
for dtype in [np.int8, np.int32, np.int64]:
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)

-128
127
-2147483648
2147483647
-9223372036854775808
9223372036854775807


`코드설명`: 정수형 자료형에 허용되는 비트수에 따른 최소값과 최대값

> 8비트 정수형 (np.int8): -128에서 127 사이의 정수
> 16비트 정수형 (np.int32): -2147483648에서 2147483647 사이의 정수
> 32비트 정수형 (np.int32): -9223372036854775808에서 9223372036854775808 사이의 정수

In [56]:
for dtype in [np.float32, np.float64]:
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
    print(np.finfo(dtype).eps)  # eps: 앱실론, 매우 작은 값

-3.4028235e+38
3.4028235e+38
1.1920929e-07
-1.7976931348623157e+308
1.7976931348623157e+308
2.220446049250313e-16


`코드설명`: 부동소수점 자료형에 허용되는 비트수에 따른 최소값과 최대값

**문제 2**

부동소수점으로 구성된 1차원 어레이가 아래와 같이 주어졌다.

In [57]:
Z = (np.random.rand(10)*100).astype(np.float32)
Z

array([21.603819 , 61.958794 , 65.02011  , 15.202485 ,  6.1349626,
       78.07616  , 45.98004  ,  5.8163795, 99.48663  ,  5.778056 ],
      dtype=float32)

변수 `Y`가 변수 `Z`가 가리키는 1차원 어레이의 dtype이 `int32`로 형변환된 1차원 어레이를 가리키도록 하라.
단, `astype()` 메서드를 이용한다.

In [58]:
# None을 적절한 표현식으로 대체하라.
Y = Z.astype(np.int32)
Y

array([21, 61, 65, 15,  6, 78, 45,  5, 99,  5], dtype=int32)

아래 코드는 수정하지 마세요!!!

In [59]:
assert (Y.shape == (10,))
assert (Y.dtype == np.int32)

**문제 3**

변수 `Z`가 아래와 같이 무작위로 생성된 길이가 10인 1차원 어레이를 가리킨다.

In [60]:
Z = np.random.uniform(0,10,10)
Z

array([6.95035229, 9.83678925, 2.39180104, 1.42249369, 1.21384939,
       3.03275148, 1.01045811, 6.92161342, 0.62291798, 5.09422128])

(1) 나머지 연산자(`%`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [61]:
# None을 적절한 표현식으로 대체하라.
Y = Z - Z%1
Y

array([6., 9., 2., 1., 1., 3., 1., 6., 0., 5.])

아래 코드는 수정하지 마세요!!!

In [62]:
assert (Y == Z.astype(np.int32)).all()

(2) 몫 연산자(`//`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [63]:
# None을 적절한 표현식으로 대체하라.
Y = Z//1

아래 코드는 수정하지 마세요!!!

In [64]:
assert (Y == Z.astype(np.int32)).all()

(3) 내림 함수 `np.floor()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [65]:
# None을 적절한 표현식으로 대체하라.
Y = np.floor(Z)

아래 코드는 수정하지 마세요!!!

In [66]:
assert (Y == Z.astype(np.int32)).all()

(4) `np.trunc()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [67]:
# None을 적절한 표현식으로 대체하라.
Y = np.trunc(Z)

아래 코드는 수정하지 마세요!!!

In [68]:
assert (Y == Z.astype(np.int32)).all()